File: #0129.sum-root-to-leaf-numbers.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int sumNumbers(TreeNode* root) {

        calcSumRootToLeafNodes(root, 0);
        return sum;
    }
    
private:
    int sum = 0;
    void calcSumRootToLeafNodes(TreeNode *root, int num) {

        if (!root) {
            return;
        }

        if (!root->left && !root->right) {
            sum += num * 10 + root->val;
            return;
        }

        calcSumRootToLeafNodes(root->left, num * 10 + root->val);
        calcSumRootToLeafNodes(root->right, num * 10 + root->val);
    }
};

File: #0129.sum-root-to-leaf-numbers.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var sumNumbers = function(root) {
    let sumNumbers = (node, sum) => {
        if (!node.left && !node.right) {
            total += sum * 10 + node.val;
        } else {
            if (node.left) {
                sumNumbers(node.left, sum * 10 + node.val);
            }
            if (node.right) {
                sumNumbers(node.right, sum * 10 + node.val);
            }
        }
    }

    let total = 0;
    sumNumbers(root, 0);
    return total;
};
